### ISSUE TITLE
Design Strategy to access from multiple IPs to prevent blocking

### ISSUE THREAD

**Issue by swathi-ssunder** [PokemonGoers/PokeData#15]

**Comment by jonas-he** [comment]
First of all i would check if sites such as pokeradar.io block you for excessively querying their service.
If yes, then we could try to send our requests through a list of proxies and cycle through them.
Will get back to it once i tested the extraction script.

**Comment by jonas-he** [comment]
As of now after ~20 seconds of continuously scraping pokeradar my IP gets blocked (no responses received -> timeout) for about 1 minute, after that it works again.
So I came up with this strategy: at the start i collect a list of free proxies (with this module https://github.com/chill117/proxy-lists) then i filter that for working proxies (with https://github.com/scastiel/node-proxy-checker). While that happens my script scrapes pokeradar for 20 secs without proxy, then on receiving a timeout (i set it to 5 secs) i switch to the first proxy and so on.

However I'm struggling with getting good proxies that are free, fast and reliable ... maybe they don't exist :smile: maybe you can help with university resources ? basically we only need the server where the app is running + 3 IPs, so they can fill the minute where the server is blocked. @sacdallago @gyachdav 

Basically this issue is just about scan speed. Without proxy a global scan of pokeradar takes 1-2 hours and with good proxies it would be a fourth of that time.
I'm also trying to optimize the scan to search only in areas where pokemon can be expected and pokeradar has found pokemon previously (which  would reduce time drastically).

**Comment by gyachdav** [comment]
I installed [simple proxy](http://benalman.com/code/projects/php-simple-proxy/docs/files/ba-simple-proxy-php.html) on my account. you can try:

https://rostlab.org/~gyachdav/proxy/ba-simple-proxy.php
http://131.159.28.140/~gyachdav/proxy/ba-simple-proxy.php

I may be able to get to more hosts but need more time. @sacdallago has access to at least one more host (got.show).

**Comment by jonas-he** [comment]
Ok thank you, works fine.

**Comment by jonas-he** [comment]
ok tested e.g.:

http://131.159.28.140/~gyachdav/proxy/ba-simple-proxy.php?url=http%3A%2F%2Fwww.pokeradar.io%2Fapi%2Fv1%2Fsubmissions%3FminLatitude%3D33.969986180360344%26maxLatitude%3D34.016455319170184%26minLongitude%3D-118.4941577911377%26maxLongitude%3D-118.46025466918944%26pokemonId%3D0

which works fine. But my script does send about 100 of such requests per second and it seems like your proxy can't handle that because i start getting responses ~half a minute after i started "flooding" your server... I also noticed that rostlab.org was not working properly after that so sorry for crashing your server :smile: @gyachdav

**Comment by gyachdav** [comment]
the rostlab.org host seems to be fine now, but if you noticed such behavior then it seems it wont be able to handle a high response load. Better if we test the capacity of the host first before we continue otherwise we will be DoSing our own servers. 💣 

@jonas-he have you tried ToR proxies?

@sacdallago, what do you suggest here?

**Comment by jonas-he** [comment]
@gyachdav yes, for now i wont use your proxy anymore. Regarding using the tor network, this would of course be possible but then we would have to configure the tor software on the server and so on, then send the requests through our local SOCKS5 proxy. Maybe this is a bit overkill, especially with all the encryption going on ...
I also made improvements with the scanning technique: Now I implented 2 versions of the scan:
- global: scan the whole world, divide the search space of -180 to 180° longitude and -90 to 90° latitude into 5°x 5° boxes and scan them one after another (takes 1-2 hours, could be reduced with proxies).
- optimized: use the results of the global scan and scan only the boxes where we actually found pokemon. This drastically reduces search space, because it turns out that only ~160 of the 1872 possible boxes contain pokemon. Takes around 5-10 minutes. But you have the possibility to miss some pokemon, if they appear in boxes where no pokemon was before in the global scan. But I think the assumption that there won't be many "misses" is pretty correct, since pokeRadar only focuses on populated regions in America/Europe and this won't change much.

So as a conclusion I came up with this idea: run one global scan a day. Then unify the set of boxes where we found at least one pokemon with the results of the previous global scans. In the mean time run the optimized scan based on this set of boxes in an infinite loop.

**Comment by sacdallago** [comment]
Let's speak about potential:
I have another 49 unused LRZ public ip's I can potentially open and another 250 internal MWN ip's I can dispose of.
The two open ones from LRZ for now are the js.dallago.us and cell.dallago.us . In the US I have christian.dallago.us and in germany again (under another provider) I have amadeus.zucchitours.com and remote.zucchitours.com.
Needless to say, I'm under strict surveillance on each and every one of them, so there's only so much traffic I can perform before I get some fancy emails from LRZ, Arvixe or HostEurope.
So, yeah: I _can_ dispose you with some proxies but it needs to be EXTREMELY under the radar

**Edit:**
@gyachdav you'd need to teach me how / you'd have to set up that proxy thing you posted up there. Also: it would need to be secure (I don't want proxy requests from random users on the web). If you give me your id_rsa.pub I can set up a few ubuntu machines in LRZ and give you ssh access.

**Comment by jonas-he** [comment]
@sacdallago what do you mean by "under the radar"? Should the traffic be rather low? The script consumes about 1 Mbps up and downstream each.
For the proxy authentication there is something like a basic HTTP authentication (which is of course not safe since it is not encrypted) or for apache you have a "Require ip" directive which only allows access from certain IP addresses. That would be an option if the IP of our server doesn't change.

**Comment by sacdallago** [comment]
@jonas-he I can go with the Require-ip directive, actually I can also just implement it in my ip-tables!

1Mbps bursts or 1Mbps constant traffic? They have quite sophisticated ways of checking you, starting with the obvious # of requests to the same host +duration of request + interval between requests.

We can give it a try anyway. I just need to be told how to set these boxes up, I would start with two (that's what we need for now, right?)

**Comment by jonas-he** [comment]
@sacdallago 1Mbps constant for about 20 seconds then it drops to zero because pokeradar blocks me -> switch to next proxy and so on.

I dont quite get what you are worried about? Will you run into traffic limitations or do you fear that the server providers will allege us of DoS'ing pokeradar? Or are you not allowed to act as a proxy?

Three would be better :smile: 
If youre already running apache on these servers, I think it would be the easiest to configure apache to act as a proxy. If not you can use the php script guy used http://benalman.com/projects/php-simple-proxy/

**Comment by sacdallago** [comment]
My problem is being labeled as DDoser :P nothing else!

Oh god, Apache. I feel like I'm back in the 90's :older_man:  ahahahah
I'll look into building up those machines soon enough, with that script... 3 it will be.

**Comment by jonas-he** [comment]
thanks for your effort! I think for a real DoS attack, my script doesn't produce enough traffic. However i can implement slight timeouts in between the requests if you want (also if you get a "fancy" email).

**Comment by jonas-he** [comment]
another thing: when you configure iptables, can you add my IP to the whitelist? So i can test from home.

**Comment by jonas-he** [comment]
any news @sacdallago ?

**Comment by sacdallago** [comment]
@jonas-he no, sorry :( it's part me, part other people :D I'll try to solve the me part asap

**Comment by gyachdav** [comment]
why is this still open?

**Comment by sacdallago** [comment]
Because it never happened, because time and uncertanty.

It's not really needed anyway, and given that I am a bit overloaded right now, I'll say we can do this in the future, if ever

**Comment by jonas-he** [comment]
i agree
